x/gamm/pool-models/balancer/amm_test.go

```diff
@@ -436,3 +436,683 @@ func TestCalcJoinPoolShares(t *testing.T) {
+		})
+	}
+}
+
+func TestCalcJoinSingleAssetTokensIn(t *testing.T) {
+	testCases := []struct {
+		name           string
+		swapFee        sdk.Dec
+		poolAssets     []balancer.PoolAsset
+		tokensIn       sdk.Coins
+		expectShares   sdk.Int
+		expectLiqudity sdk.Coins
+	}{
+		{
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t / B_t))^W_t - 1)
+			//
+			// 2_499_999_968_750 = 100 * 10^18 * (( 1 + (50,000 / 1_000_000_000_000))^0.5 - 1)
+			//
+			// where:
+			// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+			//	A_t = amount of deposited asset = 50,000
+			//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+			//	W_t = normalized weight of deposited asset in pool = 0.5 (equally weighted two-asset pool)
+			// Plugging all of this in, we get:
+			// 	Full solution: https://www.wolframalpha.com/input?i=100000000000000000000*%28%281+%2B+%2850000%2F1000000000000%29%29%5E0.5+-+1%29
+			// 	Simplified:  P_issued = 2,499,999,968,750
+			name:    "one token in - equal weights with zero swap fee",
+			swapFee: sdk.MustNewDecFromStr("0"),
+			poolAssets: []balancer.PoolAsset{
+				{
+					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+				{
+					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+			},
+			tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
+			expectShares: sdk.NewInt(2_499_999_968_750),
+		},
+		{
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t / B_t))^W_t - 1)
+			//
+			// 2_499_999_968_750 = 100 * 10^18 * (( 1 + (50,000 / 1_000_000_000_000))^0.5 - 1)
+			//
+			// where:
+			// 	P_supply = initial pool supply = 100 * 10^18 (set at pool creation, same for all new pools)
+			//	A_t = amount of deposited asset = 50,000
+			//	B_t = existing balance of deposited asset in the pool prior to deposit = 1,000,000,000,000
+			//	W_t = normalized weight of deposited asset in pool = 0.5 (equally weighted two-asset pool)
+			// Plugging all of this in, we get:
+			// 	Full solution: https://www.wolframalpha.com/input?i=100000000000000000000*%28%281+%2B+%2850000%2F1000000000000%29%29%5E0.5+-+1%29
+			// 	Simplified:  P_issued = 2,499,999,968,750
+			name:    "two tokens in - equal weights with zero swap fee",
+			swapFee: sdk.MustNewDecFromStr("0"),
+			poolAssets: []balancer.PoolAsset{
+				{
+					Token:  sdk.NewInt64Coin("uosmo", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+				{
+					Token:  sdk.NewInt64Coin("uatom", 1_000_000_000_000),
+					Weight: sdk.NewInt(100),
+				},
+			},
+			tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000), sdk.NewInt64Coin("uatom", 50_000)),
```

Comment:
Wait is this multi asset deposit test? Does it relate to #1713?

Reply:
It does not. After we do multi-asset maximal equal ratio join in`CalcJoinPoolShares`, there might be coins remaining if the share ratios [are not equal](https://github.com/osmosis-labs/osmosis/blob/dbefea1c6a4af4cbfd2e23d9d1dc8941e64e87d0/x/gamm/pool-models/internal/cfmm_common/lp.go#L82)

So we perform a single asset join on all remaining coins

Reply:
Technically, what multi-asset join in #1713 does is:
1. [maximal equal ration join](https://github.com/osmosis-labs/osmosis/blob/dbefea1c6a4af4cbfd2e23d9d1dc8941e64e87d0/x/gamm/pool-models/balancer/amm.go#L286) (it might leave some remaining tokens unjoined)
2. If any tokensIn remain, do a [single asset join on each individually](https://github.com/osmosis-labs/osmosis/blob/dbefea1c6a4af4cbfd2e23d9d1dc8941e64e87d0/x/gamm/pool-models/balancer/amm.go#L298-L313).

So, technically, this test is a step 2 of #1713 only